src/Compilers/CSharp/Test/Emit2/Semantics/LockTests.cs

```diff
@@ -0,0 +1,1102 @@
+        // Note: no warnings here as we don't expect `Lock` to be unsealed,
+        // so this doesn't warrant a warning in spec and implementation.
+        verifier.VerifyDiagnostics();
+    }
+
+    [Fact]
+    public void Downcast()
+    {
+        var source = """
+            using System;
+            using System.Threading;
+
+            object o = new Lock();
+            lock ((Lock)o) { Console.Write("L"); }
+            """;
+        var verifier = CompileAndVerify([source, LockTypeDefinition], verify: Verification.FailsILVerify,
+           expectedOutput: "ELD");
+        verifier.VerifyDiagnostics(
+            // 0.cs(4,12): warning CS9214: A value of type 'System.Threading.Lock' converted to a different type will use likely unintended monitor-based locking in 'lock' statement.
+            // object o = new Lock();
+            Diagnostic(ErrorCode.WRN_ConvertingLock, "new Lock()").WithLocation(4, 12));
+    }
+
+    [Fact]
+    public void OtherConversions()
+    {
+        var source = """
+            #nullable enable
+            using System.Threading;
+
+            I i = new Lock();
+            Lock? l = new Lock();
+            C c = new Lock();
+            D d = new Lock();
+            d = (D)new Lock();
+
+            interface I { }
+
+            class C
+            {
+                public static implicit operator C(Lock l) => new C();
+            }
+
+            class D
+            {
+                public static explicit operator D(Lock l) => new D();
+            }
+            """;
+        // No warnings about converting `Lock` expected.
+        CreateCompilation([source, LockTypeDefinition]).VerifyEmitDiagnostics(
+            // 0.cs(4,7): error CS0266: Cannot implicitly convert type 'System.Threading.Lock' to 'I'. An explicit conversion exists (are you missing a cast?)
+            // I i = new Lock();
+            Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "new Lock()").WithArguments("System.Threading.Lock", "I").WithLocation(4, 7),
+            // 0.cs(7,7): error CS0266: Cannot implicitly convert type 'System.Threading.Lock' to 'D'. An explicit conversion exists (are you missing a cast?)
+            // D d = new Lock();
+            Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "new Lock()").WithArguments("System.Threading.Lock", "D").WithLocation(7, 7));
+    }
+
+    [Theory]
+    [InlineData("")]
+    [InlineData("where T : class")]
+    [InlineData("where T : Lock")]
+    public void GenericParameter(string constraint)
+    {
+        var source = $$"""
+            using System;
+            using System.Threading;
+
+            M(new Lock());
+
+            static void M<T>(T t) {{constraint}}
+            {
+                lock (t) { Console.Write("L"); }
+            }
+            """;
+        var verifier = CompileAndVerify([source, LockTypeDefinition], verify: Verification.FailsILVerify,
+           expectedOutput: "L");
+        verifier.VerifyDiagnostics();
```

Comment:
>verifier.VerifyDiagnostics(); [](http://example.com/codeflow?start=8&length=29)

How do we know what lowering strategy was used?

Reply:
The `Lock` would output `E` at enter and `D` at dispose - we got only `L`, so `Monitor` was used. But I guess you are suggesting we should verify IL, as well?

Reply:
> But I guess you are suggesting we should verify IL, as well?

Perhaps the current output tells us that "Lock" lowering was not used, but it doesn't really tall us if Monitor lowering was used. In case of a bug we could completely ignore the `lock` part of the statement